src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -779,13 +779,47 @@ public SolutionCompilationState WithDocumentAttributes<TArg>(
             SolutionState.WithDocumentAttributes(documentId, arg, updateAttributes), documentId);
     }
 
+    /// <summary>
+    /// Creates a new solution instance with the documents specified updated to have the texts specified.
+    /// </summary>
+    /// <remarks>
+    /// This method supports updating ordinary documents and source generated documents.
+    /// </remarks>
     internal SolutionCompilationState WithDocumentTexts(ImmutableArray<(DocumentId documentId, SourceText text)> texts, PreservationMode mode)
-        => UpdateDocumentsInMultipleProjects<DocumentState, SourceText, PreservationMode>(
-            texts,
+    {
+        var (ordinaryDocuments, sourceGeneratedDocuments) = GetOrdinaryAndSourceGeneratedDocuments();
+
+        return UpdateDocumentsInMultipleProjects<DocumentState, SourceText, PreservationMode>(
+            ordinaryDocuments,
+            sourceGeneratedDocuments,
             arg: mode,
             updateDocument: static (oldDocumentState, text, mode) =>
                 SourceTextIsUnchanged(oldDocumentState, text) ? oldDocumentState : oldDocumentState.UpdateText(text, mode));
 
+        (ImmutableArray<(DocumentId, SourceText)>,
+         ImmutableArray<(SourceGeneratedDocumentIdentity, DateTime, SourceText, SyntaxNode?)>) GetOrdinaryAndSourceGeneratedDocuments()
+        {
+            if (!texts.Any(static t => t.documentId.IsSourceGenerated))
+                return (texts, []);
+
+            using var _1 = ArrayBuilder<(DocumentId, SourceText)>.GetInstance(capacity: texts.Length, out var ordinaryDocuments);
+            using var _2 = ArrayBuilder<(SourceGeneratedDocumentIdentity, DateTime, SourceText, SyntaxNode?)>.GetInstance(out var sourceGeneratedDocuments);
+            foreach (var doc in texts)
+            {
+                if (!doc.documentId.IsSourceGenerated)
+                {
+                    ordinaryDocuments.Add(doc);
+                }
+                else if (TryGetSourceGeneratedDocumentStateForAlreadyGeneratedId(doc.documentId) is { Identity: var identity })
+                {
+                    sourceGeneratedDocuments.Add((identity, DateTime.UtcNow, doc.text, null));
+                }
+            }
+
+            return (ordinaryDocuments.ToImmutableAndClear(), sourceGeneratedDocuments.ToImmutableAndClear());
+        }
+    }
```

Comment:
We should be able to extract out this helper and make it generic over (DocumentId, T) where T woudd be either SourceText or SyntaxNode, and then have it make the return array just with a few 'as' checks.

Reply:
Waiting for unions 😛